### PR TITLE
Sfpu recip sqrt implementation

### PR DESCRIPTION
**Pull request by vkrsmanovicTT**
Ticket

N/A - Initial SFPU rsqrt implementation for Quasar architecture
Problem description

The Quasar architecture needed an implementation of the rsqrt (reciprocal square root: 1/sqrt(x)) SFPU operation.
What's changed
Core Implementation:
Added ckernel_sfpu_rsqrt.h for Quasar with SFPU instructions to compute rsqrt using hardware sqrt and reciprocal operations
Implemented rsqrt by chaining SQRT_MODE and RECIP_MODE SFPU nonlinear instructions
Test Infrastructure:
Added test_sfpu_rsqrt_quasar.py Python test with random input generation in range [0.01, 1.0]
Added sfpu_rsqrt_quasar_test.cpp C++ kernel implementing datacopy + SFPU pipeline
Current Status:
Implementation is working for a limited set of format combinations (Float16 input/output)
Test sweep currently covers basic approximation mode and destination accumulation variants
Test sweep will be expanded in future updates to cover additional data formats, tile sizes, and edge cases
Type of change
[ ] Bug fix (non-breaking change which fixes an issue)
[x] New feature (non-breaking change which adds functionality)
[ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
[ ] Documentation update

### CODE REVIEW

File: tests/python_tests/helpers/test_variant_parameters.py [hunk 1]
```diff
@@ -115,11 +115,19 @@ def covert_to_cpp(self) -> str:
 
 def _generate_operation_constants(mathop: MathOperation) -> list[str]:
     """Generate the appropriate operation constants based on the math operation type."""
+    from .chip_architecture import ChipArchitecture, get_chip_architecture
+
     constants = []
 
     if mathop in SFPU_UNARY_OPERATIONS:
+        # For Quasar, use ckernel::SfpuType to avoid ambiguity with ::SfpuType from llk_sfpu_types.h
+        sfpu_type = (
```
[the remark this answers is not in the document]
Comment (vkrsmanovicTT) [reply tests/python_tests/helpers/test_variant_parameters.py]: I think that these conditionals due the fact that both enums are in scope

File: tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    if constexpr (APPROXIMATION_MODE)
+    {
+        TTI_SFPNONLINEAR(p_sfpu::LREG0, p_sfpu::LREG1, p_sfpnonlinear::SQRT_MODE); // Read value from lreg[0], approximate sqrt, load back into lreg[1]
+
+        // Then compute 1/sqrt(x) = recip(sqrt(x)) into LREG2
+        TTI_SFPNONLINEAR(p_sfpu::LREG1, p_sfpu::LREG2, p_sfpnonlinear::RECIP_MODE); // Read value from lreg[1], approximate recip, load back into lreg[2]
+    }
+
+    // Store from lreg[2] into dest register
+    TTI_SFPSTORE(p_sfpu::LREG2, 0, ADDR_MOD_7, 0, 0);
```
[the remark this answers is not in the document]
Comment (vkrsmanovicTT) [reply tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h]: Currently only APPROXIMATION MODE true is implemented and tested

File: tests/python_tests/quasar/test_sfpu_rsqrt_quasar.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    TEST_FACE_DIMS,
+    TILE_COUNT,
+    UNPACKER_ENGINE_SEL,
+)
+from helpers.utils import passed_test
+
+
+@pytest.mark.quasar
+@parametrize(
+    test_name="sfpu_rsqrt_quasar_test",
```
Comment (fvranicTT) [inline tests/python_tests/quasar/test_sfpu_rsqrt_quasar.py]:
```suggestion
```

File: tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+// Calculates RSQRT (reciprocal square root) for number of rows of output SFPU ops (Quasar = 2 rows)
+template <bool APPROXIMATION_MODE>
+inline void _calculate_rsqrt_sfp_rows_()
+{
+    TTI_SFPLOAD(p_sfpu::LREG0, p_sfpu::sfpmem::DEFAULT, ADDR_MOD_7, 0, 0); // load from dest into lreg[0], uses ADDR_MOD_7 (set to all zeroes)
+
+    // First compute sqrt(x) into LREG1
+    if constexpr (APPROXIMATION_MODE)
+    {
+        TTI_SFPNONLINEAR(p_sfpu::LREG0, p_sfpu::LREG1, p_sfpnonlinear::SQRT_MODE); // Read value from lreg[0], approximate sqrt, load back into lreg[1]
```
Comment (fvranicTT) [inline tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h]: Please include "ckernel_ops.h" for these defines.

File: tests/sources/quasar/sfpu_rsqrt_quasar_test.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+void run_kernel(const volatile struct RuntimeParams *params)
+{
+    uint32_t const buf_desc_id    = 8;
+    const uint num_tiles_per_pack = params->TILE_CNT;
+
+    // Setup dvalid for PACK
+    set_up_dest_dvalid_per_thread<dest_dvalid_client::PACK>({dest_dvalid_client::FPU, dest_dvalid_client::SFPU, dest_dvalid_client::PACK});
+
+    buffer_descriptor_u bd_val = {0};
+    tdma_descriptor_t tdma_desc;
```
[the remark this answers is not in the document]
Comment (vkrsmanovicTT) [reply tests/sources/quasar/sfpu_rsqrt_quasar_test.cpp]: Done. Thanks

File: tests/python_tests/quasar/test_sfpu_rsqrt_quasar.py (hunk 3 above)
Comment (vkrsmanovicTT) [reply tests/python_tests/quasar/test_sfpu_rsqrt_quasar.py]: Removed according to refactoring

File: tests/python_tests/quasar/test_sfpu_rsqrt_quasar.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    input_dimensions=[[32, 32], [64, 64]],
+)
+def test_sfpu_rsqrt_quasar(
+    formats, approx_mode, dest_acc, implied_math_format, input_dimensions
+):
+    """
+    Test reciprocal square root (rsqrt) operation on Quasar architecture.
+
+    Uses PyTorch's rsqrt as the golden reference and generates input stimuli
+    in the range (0.1, 2.0] to test rsqrt behavior.
```
[the remark this answers is not in the document]
Comment (rtawfik01) [reply tests/python_tests/quasar/test_sfpu_rsqrt_quasar.py]: @vkrsmanovicTT is using SFPNONLINEAR, which has higher accuracy in a smaller range.

File: tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+namespace sfpu
+{
+// Calculates RSQRT (reciprocal square root) for number of rows of output SFPU ops (Quasar = 2 rows)
+template <bool APPROXIMATION_MODE>
+inline void _calculate_rsqrt_sfp_rows_()
+{
+    TTI_SFPLOAD(p_sfpu::LREG0, p_sfpu::sfpmem::DEFAULT, ADDR_MOD_7, 0, 0); // load from dest into lreg[0]
+
+    // First compute sqrt(x) into LREG1
+    if constexpr (APPROXIMATION_MODE)
```
[the remark this answers is not in the document]
Comment (vkrsmanovicTT) [reply tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h]: I modified and ran the tests for rsqrt quasar for a range from 0.1 to 10 , negative values excluded because of sqrt. 
All the tests pass which means the pcc value is in the desired range.

File: tests/python_tests/quasar/test_sfpu_rsqrt_quasar.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    MATH_OP,
+    NUM_FACES,
+    TEST_FACE_DIMS,
+    TILE_COUNT,
+    UNPACKER_ENGINE_SEL,
+)
+from helpers.utils import passed_test
+
+
+def generate_sfpu_rsqrt_combinations(
```
[the remark this answers is not in the document]
Comment (vkrsmanovicTT) [reply tests/python_tests/quasar/test_sfpu_rsqrt_quasar.py]: Thanks for the hint

File: tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (rtawfik01) [reply tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h]: if you can test it up to full max range, and PCC still is ok, then we can remove the APPROX_MODE flag (guess I was being too pessimistic here)

File: tests/sources/quasar/sfpu_rsqrt_quasar_test.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+        _llk_math_set_dvalid_<p_cleardvalid::FPU>();
+    }
+
+    _llk_math_eltwise_unary_sfpu_init_();
+
+    // Apply SFPU rsqrt to all tiles
+    for (int i = 0; i < params->TILE_CNT; ++i)
+    {
+        _llk_math_eltwise_unary_sfpu_params_<true>(ckernel::sfpu::_calculate_rsqrt_, i, num_sfpu_iterations);
```
Comment (rtawfik01) [inline tests/sources/quasar/sfpu_rsqrt_quasar_test.cpp]: Remove the APPROX_MODE from template here too

File: tt_llk_quasar/llk_lib/llk_math_eltwise_unary_sfpu_common.h [hunk 10]
```diff
@@ -74,15 +74,12 @@ inline void _llk_math_eltwise_unary_sfpu_init_()
...
- * @tparam APPROXIMATE: Some sfpu functions have 2 implementations
- * either less accurate and more performant mode, or more accurate but less performant mode
- * APPROXIMATE flag is set for more performant but less accurate mode
  * @param: sfpu_func: function pointer to the sfpu functions to run, can look at list of functions here: common/inc/sfpu/cmath_sfpu*
  * @param: dst_tile_index: Starting tile index in the destination register, values = 0 - 15
  * @param: args: variable number of args can be passed into this function, that will be passed
  * to the SFPU function pointer
  */
-template <bool APPROXIMATE, class F, class... ARGS>
+template <class F, class... ARGS>
```
[the remark this answers is not in the document]
Comment (vkrsmanovicTT) [reply tt_llk_quasar/llk_lib/llk_math_eltwise_unary_sfpu_common.h]: Reverted

[the remark this answers is not in the document]
Comment (rtawfik01) [reply tt_llk_quasar/llk_lib/llk_math_eltwise_unary_sfpu_common.h]: ah yes my bad, only remove from rsqrt

File: tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h [hunk 11]
```diff
@@ -0,0 +1,41 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
```
Comment (fvranicTT) [inline tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h]:
```suggestion
// SPDX-FileCopyrightText: © 2026 Tenstorrent AI ULC
```

File: tests/python_tests/quasar/test_sfpu_rsqrt_quasar.py [hunk 12]
```diff
@@ -0,0 +1,211 @@
+# SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
```
Comment (fvranicTT) [inline tests/python_tests/quasar/test_sfpu_rsqrt_quasar.py]:
```suggestion
# SPDX-FileCopyrightText: © 2026 Tenstorrent AI ULC
```

**Review by fvranicTT — APPROVED** [review]

File: tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h (hunk 11 above)
Comment (vkrsmanovicTT) [reply tt_llk_quasar/common/inc/sfpu/ckernel_sfpu_rsqrt.h]: DOne
